src/compiler/parser.ts

```diff
@@ -904,28 +904,37 @@ namespace ts {
             // overhead.  This functions allows us to set all the parents, without all the expense of
             // binding.
 
-            let parent: Node = rootNode;
-            forEachChild(rootNode, visitNode);
+            const stack: [Node, Node[]][] = [[rootNode, gatherChildren(rootNode)]];

```

Comment:
I think you can do this with just a `const stack: Node[] = [rootNode]`. You can move the logic for checking/setting parents into this function and just push the children that still need to be visited onto the stack.